doc/namespace.md

```diff
@@ -0,0 +1,417 @@
+
+```ruby
+# in foo.rb
+class String
+  BLANK_PATTERN = /\A\s*\z/
+  def blank?
+    self =~ BLANK_PATTERN
+  end
+end
+
+module Foo
+  def self.foo = "foo"
+
+  def self.foo_is_blank?
+    foo.blank?
+  end
+end
+
+Foo.foo.blank? #=> false
+"foo".blank?   #=> false
+
+# in main.rb
+ns = Namespace.new
+ns.require('foo')
+
+Foo.foo_is_blank? #=> false   (#blank? called in ns)
+
+Foo.foo.blank?    # NoMethodError
+"foo".blank?      # NoMethodError
+String::BLANK_PATTERN # NameError
+```
+
+The main namespace and `ns` are different namespaces, so monkey patches in main are also invisible in `ns`.
+
+### Builtin classes and modules
+
+In the namespace context, "builtin" classes and modules are classes and modules:
+
+* Accessible without any `require` calls in user scripts
+* Defined before any user program start running
+* Including classes/modules loaded by `prelude.rb` (including RubyGems `Gem`, for example)
+
+Hereafter, "builtin classes and modules" will be referred to as just "builtin classes".
+
+### Builtin classes referred via namespace objects
+
+Builtin classes in a namespace `ns` can be referred from other namespace. For example, `ns::String` is a valid reference, and `String` and `ns::String` are identical (`String == ns::String`, `String.object_id == ns::String.object_id`).
+
+`ns::String`-like reference returns just a `String` in the current namespace, so its definition is `String` in the namespace, not in `ns`.
+
+```ruby
+# foo.rb
+class String
+  def self.foo = "foo"
+end
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+ns::String.foo  # NoMethodError
+```
+
+### Class instance variables, class variables, constants
+
+Builtin classes can have different sets of class instance variables, class variables and constants between namespaces.
+
+```ruby
+# foo.rb
+class Array
+  @v = "foo"
+  @@v = "_foo_"
+  V = "FOO"
+end
+
+Array.instance_variable_get(:@v) #=> "foo"
+Array.class_variable_get(:@@v)   #=> "_foo_"
+Array.const_get(:V)              #=> "FOO"
+
+# main.rb
+ns = Namespace.new
+ns.require('foo')
+
+Array.instance_variable_get(:@v) #=> nil
+Array.class_variable_get(:@@v)   # NameError
+Array.const_get(:V)              # NameError
+```
+
+### Global variables
+
+In namespaces, changes on global variables are also isolated in the namespace. Changes on global variables in a namespace are visible/applied only in the namespace.
+
+```ruby
+# foo.rb
+$foo = "foo"
+$VERBOSE = nil
+
+puts "This appears: '#{$foo}'"
+
+# main.rb
+p $foo      #=> nil
+p $VERBOSE  #=> false
+
+ns = Namespace.new
+ns.require('foo')  # "This appears: 'foo'"
+
+p $foo      #=> nil
+p $VERBOSE  #=> false
+```
+
+### Top level constants
+
+Usually, top level constants are defined as constants of `Object`. In namespaces, top level constants are constants of `Object` in the namespace. And the namespace object `ns`'s constants are strictly equal to constants of `Object`.
```

Comment:
> And the namespace object `ns`'s constants are strictly equal to constants of `Object`.

How is this implemented? Is the constants table shared between both the Namespace and that namespace's Object?
That seems quite dangerous implementation-wise as it might have far-reaching implications on constant lookup caching and invalidation.
I think it would be better to preserve the invariant that a constant table is only used for one Module, and implement `ns::Foo` another way, e.g. via `const_missing`.

Reply:
> How is this implemented? Is the constants table shared between both the Namespace and that namespace's Object?

Correct.
As far as I checked (and CI checked), it's working correctly. But I'll rethink if we find it'll not work well.